### PR TITLE
New ui element for comments in config (closes #2348)

### LINKED ISSUE #2348: Add new ui element to generate valid json for motion comment field config
The motion config "Comment fields for motions" required valid json if user wants to add private fields, e.g. `[{"name": "foo", "public": false},{"name": "bar", "public": true}]`.

This issue should help the user to generate such json string automatically. Suggestion: User just enters comment fields comma separated in the input field. All fields are listed below the input with a checkbox for each field to switch it to private (default: public).

### PR DESCRIPTION
**Pull request by FinnStutzenstein**
A new approve to store the comments in the config: A list with objects is now saved rather than a JSON string. For this reason the complete JSON validation could is removed and not error prone anymore. Additional the new ui element makes it simple to create, modify and delete comments.

### CODE REVIEW

File: openslides/motions/access_permissions.py [hunk 1]
```diff
@@ -34,7 +30,7 @@ def get_restricted_data(self, full_data, user):
             data = full_data
         else:
             data = full_data.copy()
-            for i, field in enumerate(self.get_comments_config_fields()):
+            for i, field in enumerate(config['motions_comments']):
```
Comment (emanuelschuetze) [inline openslides/motions/access_permissions.py]: We have to check here if this config value is a list of dictionaries.

File: openslides/motions/access_permissions.py [hunk 2]
```diff
@@ -1,7 +1,3 @@
-import json
-
-from jsonschema import ValidationError, validate
```
Comment (emanuelschuetze) [inline openslides/motions/access_permissions.py]: You have to remove jsonschema also from requirements_production.txt.

File: bower.json [hunk 3]
```diff
@@ -25,7 +25,6 @@
     "jquery.cookie": "~1.4.1",
     "js-data": "~2.8.2",
     "js-data-angular": "~3.1.0",
-    "jsen": "~0.6.1",
```
Comment (emanuelschuetze) [inline bower.json]: Remove jsen also from README.rst (line 219).

File: openslides/motions/static/js/motions/site.js [hunk 4]
```diff
@@ -1054,7 +1054,7 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions', 'OpenSlid
         Motion.loadRelations(motion, 'agenda_item');
         $scope.version = motion.active_version;
         $scope.isCollapsed = true;
-        $scope.commentsFields = MotionComment.getFields();
+        $scope.commentsFields = $scope.config('motions_comments');
```
Comment (emanuelschuetze) [inline openslides/motions/static/js/motions/site.js]: Why not use `Config.get('motions_comments').value`?

**Comment by emanuelschuetze** [comment]
Very nice ui element! Some remarks.

File: openslides/motions/static/js/motions/site.js (hunk 4 above)
Comment (FinnStutzenstein) [inline openslides/motions/static/js/motions/site.js]: I was thinking at the discussion in PR #2352. I will test `Config.get`.

**Comment by FinnStutzenstein** [comment]
@emanuelschuetze Ready and rebased

File: openslides/motions/access_permissions.py (hunk 1 above)
Comment (emanuelschuetze) [inline openslides/motions/access_permissions.py]: Ok.

File: openslides/motions/access_permissions.py (hunk 2 above)
Comment (emanuelschuetze) [inline openslides/motions/access_permissions.py]: Ok, done.

File: bower.json (hunk 3 above)
Comment (emanuelschuetze) [inline bower.json]: Ok (checked). Done.

File: openslides/motions/static/js/motions/site.js (hunk 4 above)
Comment (emanuelschuetze) [inline openslides/motions/static/js/motions/site.js]: If you use Config.get(..) `MotionComment` provider is unused in MotionDetailCtrl and can be removed. @normanjaeckel Is it ok for you to use Config.get(..) here?

**Comment by emanuelschuetze** [comment]
@normanjaeckel Ready for review. One remark with open question for you.

File: openslides/core/config.py [hunk 5]
```diff
@@ -98,6 +99,15 @@ def __setitem__(self, key, value):
                     value['height'] < 600 or value['height'] > 2160):
                 raise ConfigError(_('The Resolution have to be between 800x600 and 3840x2160.'))
 
+        if config_variable.input_type == 'comments':
+            for comment in value:
```
Comment (normanjaeckel) [inline openslides/core/config.py]: It is not checked that an element of the list is a dict.

File: openslides/motions/access_permissions.py [hunk 6]
```diff
@@ -34,7 +31,9 @@ def get_restricted_data(self, full_data, user):
             data = full_data
         else:
             data = full_data.copy()
-            for i, field in enumerate(self.get_comments_config_fields()):
+            if not isinstance(config['motions_comments'], list) or not all(isinstance(c, dict) for c in config['motions_comments']):
+                raise ValidationError({'detail': 'motions_comments has to be a list of dictionaries.'})
```
Comment (normanjaeckel) [inline openslides/motions/access_permissions.py]: This method should not run any validation but remove irregular fields from output or set output completely to None.

File: openslides/motions/static/js/motions/base.js [hunk 7]
```diff
@@ -310,79 +310,17 @@ angular.module('OpenSlidesApp.motions', [
...
-                                'public': true
-                            };
-                        }
-                    );
-                }
-                return fields;
-            },
             getFormFields: function () {
-                var fields = this.getFields();
+                var fields = Config.get('motions_comments').value;
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/base.js]: This check is not necessary. A non authorized user is not allowed to edit any comment field.

**Comment by FinnStutzenstein** [comment]
@normanjaeckel Ping.

File: openslides/core/config.py [hunk 8]
```diff
@@ -98,6 +99,19 @@ def __setitem__(self, key, value):
...
 
+        if config_variable.input_type == 'comments':
+            if not isinstance(value, list):
+                raise ConfigError(_('motions_comments has to be a list.'))
+            for comment in value:
+                if not isinstance(comment, dict):
+                    raise ConfigError(_('Each element in motions_comments has to be a dict.'))
+                if comment.get('name') is None or comment.get('public') is None:
+                    raise ConfigError(_('A name and a public property have to be given.'))
+            if not isinstance(comment['name'], str):
```
Comment (normanjaeckel) [inline openslides/core/config.py]: Indent the next lines. They have to be inside the loop.

Comment (normanjaeckel) [inline openslides/core/config.py]: A unit test or an integration test would be nice here. :-)

File: openslides/motions/access_permissions.py [hunk 9]
```diff
@@ -34,13 +30,19 @@ def get_restricted_data(self, full_data, user):
             data = full_data
         else:
             data = full_data.copy()
-            for i, field in enumerate(self.get_comments_config_fields()):
-                if not field.get('public'):
-                    try:
+            if isinstance(config['motions_comments'], list):
```
Comment (normanjaeckel) [inline openslides/motions/access_permissions.py]: You do not have to check the integrity of the config value here any more. You checked it already at input time.

File: openslides/motions/access_permissions.py [hunk 10]
```diff
@@ -34,13 +30,19 @@ def get_restricted_data(self, full_data, user):
...
+                for i, field in enumerate(config['motions_comments']):
+                    if not isinstance(field, dict):
```
Comment (normanjaeckel) [inline openslides/motions/access_permissions.py]: Here the same.

File: tests/integration/motions/test_viewset.py [hunk 11]
```diff
@@ -106,6 +106,22 @@ def test_with_tag(self):
...
+            { 'name': 'comment1', 'public': True},
+            { 'name': 'comment2', 'public': False}]
+        comments = ['comemnt1_sdpoiuffo3%7dwDwW&', 'comment2_iusd&D/TdskDWH&5DWas46WAd078']
+        response = self.client.post(
+            reverse('motion-list'),
+            {'title': 'title_test_sfdAaufd56HR7sd5FDq7av',
+             'text': 'text_test_fiuhefF86()ew1Ef346AF6W',
+             'comments': comments})
+        for data in response:
+            print(repr(data))
```
Comment (FinnStutzenstein) [inline tests/integration/motions/test_viewset.py]: @normanjaeckel Why there is this error: `Data must ba an array.`? I send an array, but the serializer gets a string with the last comment..

Comment (normanjaeckel) [inline tests/integration/motions/test_viewset.py]: I think you have tu submit a kind of format in the post call. See https://github.com/OpenSlides/OpenSlides/blob/master/tests/integration/agenda/test_views.py#L31

File: tests/integration/motions/test_viewset.py [hunk 12]
```diff
@@ -113,9 +113,13 @@ def test_with_multiple_comments(self):
...
             reverse('motion-list'),
-            {'title': 'title_test_sfdAaufd56HR7sd5FDq7av',
-             'text': 'text_test_fiuhefF86()ew1Ef346AF6W',
-             'comments': comments})
+            {
+                'title': 'title_test_sfdAaufd56HR7sd5FDq7av',
+                'text': 'text_test_fiuhefF86()ew1Ef346AF6W',
+                'comments': comments,
+            },
+            format='json',
```
Comment (FinnStutzenstein) [inline tests/integration/motions/test_viewset.py]: @normanjaeckel No, this doesn't work, but i do not know why. I mean it should be working..

File: openslides/motions/access_permissions.py (hunk 9 above)
Comment (FinnStutzenstein) [inline openslides/motions/access_permissions.py]: Done.

File: openslides/motions/access_permissions.py (hunk 10 above)
Comment (FinnStutzenstein) [inline openslides/motions/access_permissions.py]: Done.

File: tests/integration/motions/test_viewset.py (hunk 12 above)
Comment (normanjaeckel) [inline tests/integration/motions/test_viewset.py]: Use self.client = APIClient(), see the complete test (including setUp()) I linked in last comment.

**Comment by FinnStutzenstein** [comment]
@normanjaeckel Ping.

File: openslides/motions/access_permissions.py [hunk 13]
```diff
@@ -34,13 +30,16 @@ def get_restricted_data(self, full_data, user):
...
             data = full_data.copy()
-            for i, field in enumerate(self.get_comments_config_fields()):
+            for i, field in enumerate(config['motions_comments']):
                 if not field.get('public'):
                     try:
                         data['comments'][i] = None
                     except IndexError:
                         # No data in range. Just do nothing.
                         pass
+            else:
```
Comment (normanjaeckel) [inline openslides/motions/access_permissions.py]: This else clause is useless.

Comment (FinnStutzenstein) [inline openslides/motions/access_permissions.py]: Done. Sorry, I have overlooked it..

**Comment by FinnStutzenstein** [comment]
@normanjaeckel ok

**Comment by emanuelschuetze** [comment]
Ok, successfully retested. Merge...

**Pull request merged**